### PR TITLE
feat: store account metadata in portfolioSlice

### LINKED ISSUE #2546: Store account meta data in portfolioSliceCommon
### Overview

store BIP44 params associated with accountSpecifier when it's created for reuse across the app
https://github.com/shapeshift/web/blob/develop/src/state/slices/portfolioSlice/portfolioSliceCommon.ts#L89

See [here](https://github.com/shapeshift/web/compare/account-number-shenanigans#diff-3316dabcd3452fc0a6688adad4d102c689cc2d52d71ce1d79196e92054f630b0R26) for one potential strategy and notes

### PR DESCRIPTION
**Pull request by asamere**
## Description
depends on https://github.com/shapeshift/lib/pull/992

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #2546 

### Engineering

See inline comments below

* adds a new `accountMetadataById` section to the `portfolioSlice`
* adds some selectors we'll need extensively throughout the app (currently unused)

### Operations

* no user facing changes here, check that the general wallet connecting/switching/reconnection flow work as expected

### CODE REVIEW

**Comment by 0xdef1cafe** [comment]
image

lookin hot locally linked to the lib PR

File: src/state/slices/portfolioSlice/portfolioSliceCommon.ts [hunk 1]
```diff
@@ -77,12 +77,42 @@ export type PortfolioAccountBalances = {
...
+ * * Bitcoin Segwit Native this is a bc1-prefixed address
+ *
+ * as soon as we change websocket subscription logic, this entire byId section
+ * can be removed from the store, as the only reason we currently maintain this mapping
+ * is to be able to lookup an account from an address for websocket messages for UTXO chains
+ */
+type Address = string
+
 export type PortfolioAccountSpecifiers = {
+  accountMetadataById: AccountMetadataById
```
Comment (0xdef1cafe) [inline src/state/slices/portfolioSlice/portfolioSliceCommon.ts:110]: this is the guts of the PR - a map keyed by an `AccountId` revealing a required `BIP44Params` and optional `accountType`

we will use this throughout future PRs to be able to retrieve the `accountNumber` (and optionally `accountType` for UTXO chains)

File: src/state/slices/portfolioSlice/portfolioSliceCommon.ts [hunk 2]
```diff
@@ -77,12 +77,42 @@ export type PortfolioAccountBalances = {
...
   byId: {
     // this maps an account identifier to a list of addresses
     // in the case of utxo chains, an account (e.g. xpub/ypub/zpub) can have multiple addresses
     // in account based chains, this is a 1:1 mapping, i.e. the account is the address
-    [k: AccountSpecifier]: AccountId[]
+    [k: AccountSpecifier]: Address[]
```
Comment (0xdef1cafe) [inline src/state/slices/portfolioSlice/portfolioSliceCommon.ts:115]: this was previously wrong, clarified for now, and will be removed entirely in a future PR.

File: src/state/slices/portfolioSlice/portfolioSlice.ts [hunk 3]
```diff
@@ -23,6 +23,10 @@ export const portfolio = createSlice({
       moduleLogger.info('clearing portfolio')
       return initialState
     },
+    setAccountMetadata: (state, { payload }: { payload: AccountMetadataById }) => {
```
Comment (0xdef1cafe) [inline src/state/slices/portfolioSlice/portfolioSlice.ts:26]: putting this here in the `portfolioSlice` over the `accountSpecifiers` slice as we want to kill `accountSpecifiers` slice as it's redundant and confusing

File: src/context/AppProvider/AppContext.tsx [hunk 4]
```diff
@@ -267,7 +278,7 @@ export const AppProvider = ({ children }: { children: React.ReactNode }) => {
         switch (chainId) {
           case cosmosChainId:
           case osmosisChainId:
-            const accountSpecifier = `${chainId}:${account}`
+            const accountSpecifier = toAccountId({ chainId, account })
```
Comment (gomesalexandre) [inline src/context/AppProvider/AppContext.tsx:281]: 💜

**Review by gomesalexandre — APPROVED** [review]
Conceptually happy with this one as:

- It brings a new metadata structure over and doesn't break anything
- Conforms to our idea of a join between slices
- Selectors look sane and aren't consumed yet
- This starts solving our `accountId` / `accountSpecifier` terminology with the `Address` type
- This is the starting point we need to start going away from `accounts[0]`

Tested it in redux debugger with a native wallet with ETH/UTXOs/Atom/Osmosis accounts and the bip44Params / accountType properties look good.

[cool](https://app.graphite.dev/macros/shapeshift)

**Pull request merged**